pennylane_lightning/core/src/simulators/lightning_gpu/utils/MPIManager.hpp

```diff
@@ -0,0 +1,815 @@
+        MPI_Status status;
+        const int tag = MPI_ANY_TAG;
+
+        PL_MPI_IS_SUCCESS(MPI_Recv(recvBuf.data(), recvBuf.size(), datatype,
+                                   static_cast<int>(source), tag,
+                                   this->getComm(), &status));
+    }
+
+    /**
+     * @brief MPI_Sendrecv wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer.
+     * @param dest Rank of destination.
+     * @param recvBuf Receive buffer.
+     * @param source Rank of source.
+     */
+    template <typename T>
+    void Sendrecv(T &sendBuf, size_t dest, T &recvBuf, size_t source) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Status status;
+        int sendtag = 0;
+        int recvtag = 0;
+        int destInt = static_cast<int>(dest);
+        int sourceInt = static_cast<int>(source);
+        PL_MPI_IS_SUCCESS(MPI_Sendrecv(&sendBuf, 1, datatype, destInt, sendtag,
+                                       &recvBuf, 1, datatype, sourceInt,
+                                       recvtag, this->getComm(), &status));
+    }
+
+    /**
+     * @brief MPI_Sendrecv wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer vector.
+     * @param dest Rank of destination.
+     * @param recvBuf Receive buffer vector.
+     * @param source Rank of source.
+     */
+    template <typename T>
+    void Sendrecv(std::vector<T> &sendBuf, size_t dest, std::vector<T> &recvBuf,
+                  size_t source) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Status status;
+        int sendtag = 0;
+        int recvtag = 0;
+        int destInt = static_cast<int>(dest);
+        int sourceInt = static_cast<int>(source);
+        PL_MPI_IS_SUCCESS(MPI_Sendrecv(sendBuf.data(), sendBuf.size(), datatype,
+                                       destInt, sendtag, recvBuf.data(),
+                                       recvBuf.size(), datatype, sourceInt,
+                                       recvtag, this->getComm(), &status));
+    }
+
+    template <typename T>
+    void Scan(T &sendBuf, T &recvBuf, const std::string &op_str) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Op op = getMPIOpType(op_str);
+
+        PL_MPI_IS_SUCCESS(
+            MPI_Scan(&sendBuf, &recvBuf, 1, datatype, op, this->getComm()));
+    }
+
+    /**
+     * @brief Creates new MPIManager based on colors and keys.
+     *
+     * @param color Processes with the same color are in the same new
+     * communicator.
+     * @param key Rank assignment control.
+     * @return new MPIManager object.
+     */
+    auto split(size_t color, size_t key) -> MPIManager {
+        MPI_Comm newcomm;
+        int colorInt = static_cast<int>(color);
+        int keyInt = static_cast<int>(key);
+        PL_MPI_IS_SUCCESS(
+            MPI_Comm_split(this->getComm(), colorInt, keyInt, &newcomm));
+        return MPIManager(newcomm);
+    }
+};
```

Comment:
Given the coverage warnings introduced here, let's also create a story about improving coverage to be revisited. I think we can accept the coverage hit introduced here until then.

Reply:
Done!